### PR TITLE
Fix Some Bugs With Using Training Archives

### LINKED ISSUE #82: Problem importing previous training data when using gaussian process
**Describe the bug**
I am trying to use data from previous experiments as input for my optimizer. Unfortunately this does not work if I use a Gaussian Process controller (yet, for a Differential Evolution controller everything seems to work). Is this a known issue and is there a workaround?

**To Reproduce**
1. create a simple controller and optimize (for example the optimizer from the tutorial):

```python
#Imports for M-LOOP
import mloop.interfaces as mli
import mloop.controllers as mlc
import mloop.visualizations as mlv
import mloop.utilities as mlu

class CustomInterface(mli.Interface):
    
    def __init__(self):
        super(CustomInterface,self).__init__()
        self.minimum_params = np.array([0,0.1,-0.1])

    def get_next_cost_dict(self,params_dict):
        
        params = params_dict['params']
        
        cost = -np.sum(np.sinc(params - self.minimum_params))
        uncer = 0
        bad = False
        
        cost_dict = {'cost':cost, 'uncer':uncer, 'bad':bad}
        return cost_dict
    
def main():
    
    input_dict = mlu.get_dict_from_file('config.txt', 'txt')
    interface = CustomInterface()
    controller = mlc.create_controller(interface, **input_dict)
    controller.optimize() 

if __name__ == '__main__':
    main()
```

with the configuration file
```python
max_num_runs = 15
target_cost = -2.99
num_params = 3
min_boundary = [-2,-2,-2]
max_boundary = [2,2,2]
controller_type = 'differential_evolution'
param_names=None

training_type = 'differential_evolution'   
gp_training_filename = "training_data.txt"
gp_training_file_type = 'txt'        
```

where I either use `controller_type = 'differential_evolution'` or `controller_type = 'gaussian_process'`.  `training_data.txt` simply is a learner archive from a previous run.

**Expected behavior**

- previous run  `controller_type = 'differential_evolution'` and current run  `controller_type = 'differential_evolution'`: works
- previous run  `controller_type = 'gaussian_process'` and current run  `controller_type = 'differential_evolution'`: works
- previous run  `controller_type = 'gaussian_process'` and current run  `controller_type = 'gaussian_process'`: `TypeError: __init__() got multiple values for keyword argument 'param_names'`
- previous run  `controller_type = 'differential_evolution'` and current run  `controller_type = 'gaussian_process'`: `TypeError: __init__() got multiple values for keyword argument 'param_names'`

**Screenshots**
image

**Desktop (please complete the following information):**

- OS: Ubuntu 16.04
- M-LOOP version 3.1.1
- Python version 3.5.2

### PR DESCRIPTION
**Pull request by zakv**
Partially fixes #82, which pointed out that using `gp_training_filename`  to load data from a previous optimization's learner archive for a new optimization was broken. It turns out `nn_training_filename` had similar issues as well. It should now be possible to use old gaussian process learner archives for new optimizations with a gaussian process learner, and similarly it should be possible to use old neural net learner archives for new optimizations with the neural net learner.

This doesn't resolve all of the points in #82. In particular it is still not possible to use a learner archive from one type of learner as the training archive for another type of learner. Doing that is a bit trickier because the training archive won't have all of the properties expected by the new learner, so that would have to be handled appropriately. For example, the `GaussianProcessLearner` looks for `noise_level_history` which isn't included in archives generated by other learners. It's likely possible to set it up so that learners can take the important info from archives of other types, i.e. the costs, uncertainties, parameter values, and so on, but that has not been done yet. For now trying to use one archive type as the training archive for another learner raises a `NotImplementedError`.

Changes proposed in this pull request:

- Resolve bugs that prevented users from using the `gp_training_filename` and `nn_training_filename` arguments to load data from past learner archives for a new optimization with the same kind of learner.
- In the process, did some refactoring of code duplicated between `GaussianProcessLearner` and `NeuralNetLearner`.
  - This was done by adding an intermediate `MachineLearner` class, to be shared by both of them as an intermediate abstraction layer.
  - There's more refactoring of shared code to be done, but I plan to do it bit by bit over time.
- Added the `gp_training_override_kwargs` option to the Gaussian process learner.
  - This was necessary to allow the user to specify values for e.g. `cost_has_noise` that are different than what was used during the optimization that produced the training dataset (i.e. make keyword argument values override settings recorded in the training archive), while still allowing the `GaussianProcessVisualizer` class to load the values from the archive regardless of what keyword argument values were passed (i.e. make the training archive settings override the keyword argument values).
  - A `ValueError` is raised if some important values don't agree though, such as `num_params`.
  - The docs were updated accordingly.
- Raise a `NotImplementedError` when trying to use the learner archive from one type of learner for a new optimization with a different kind of learner.
- Fixed a bug where the `GaussianLearnerVisualizer` could have issues plotting results from an archive if its training archive used different options.
  - For example, if `cost_has_noise` was `True` in one optimization but `False` in the other, then the number of fits performed would be greater than the number of noise level fits, which would mess up the plotting. This has now been fixed.

@charmasaur and @sebbleroni, are either of you interested in taking a look at this before I merge it?

### CODE REVIEW

**Comment by charmasaur** [comment]
Thanks Zak! I'll take a look at this tomorrow.

File: mloop/learners.py [hunk 1]
```diff
@@ -1626,154 +1840,61 @@ class NeuralNetLearner(Learner, mp.Process):
...
-            self.best_index = int(self.training_dict['best_index'])
-            self.worst_cost = float(self.training_dict['worst_cost'])
-            self.worst_index = int(self.training_dict['worst_index'])
-            self.cost_range = float(self.training_dict['cost_range'])
-            
-            #Configuration of the fake neural net learner
+            # Configuration of the fake neural net learner
+            # TODO: length_scale and noise_level aren't used by neural net
+            # learner; probably a copy/paste from gaussian process learner.
+            # Maybe they should be deleted?
```
Comment (charmasaur) [inline mloop/learners.py:1861]: :+1:, definitely me copy/pasting (for all of them)

**Review by charmasaur — COMMENTED** [review]
A few completely trivial comments. I don't have the spirit to review the diff properly on a Friday evening, so I'll just trust that nothing too bad happened in the refactoring. In any case, at a high level it all seems sensible (in general I don't like using base classes just for the sake of sharing code, but given the way everything in M-LOOP is set up already, I can't really suggest anything else).

File: mloop/learners.py (hunk 1 above)
Comment (zakv) [reply mloop/learners.py:1861]: Good to know, maybe now's a good time for me to go ahead and delete them then

**Comment by zakv** [comment]
> in general I don't like using base classes just for the sake of sharing code, but given the way everything in M-LOOP is set up already, I can't really suggest anything else

Hmm, would you prefer if things were refactored so that `GaussianProcessLearner` and `NeuralNetLearner` inherit directly from `Learner` and a `MachineLearnerMixin` class, which would be used to house their shared code? I could also change that in a future PR since there's more duplicated code between those classes that I'd like to consolidate.

File: mloop/learners.py (hunk 1 above)
Comment (zakv) [reply mloop/learners.py]: Removed in 423e50d.

**Comment by charmasaur** [comment]
I don't have much of a preference between this and the mixin, I see them as pretty similar (well, given that this is already implemented, my preference is for this!).

To attempt to briefly articulate my problem with the inheritance used throughout M-LOOP, I think it's quite hard to define what the relationships are between the base and derived classes. If I'm implementing a new derived class, how do I know what is and isn't handled with the base class, and what I need to implement myself? If I'm the base class, how do I know what I'm allowed to change without breaking the derived classes? In my experience things usually end up working out more nicely (from flexibility, readability, and maintainability points of view) if the "base" class instead just accepts a dependency (or dependencies) that do the work it doesn't know how to do itself (for example that might look like fitting the model and predicting new parameters): that way you can define an explicit narrow contract for the dependency (as an abstract base class with a handful of methods), a reader doesn't need to reason about any code being overridden by other code, and a maintainer knows that the only interaction between any of the classes is via the (narrow) contract. That sort of thing would constitute a pretty major architectural shift though, which I don't think we should take lightly (because, at the end of the day, despite my old-man rambling complaints about inheritance, the code is working).

Now, to be clear, I don't think what you're doing here is at all crazy in the scheme of all things M-LOOP (and I know that I'm at least as much to blame as anyone for that scheme), since it fits in well with the design used elsewhere -- in fact I think it's quite a bit *better* than the other instances of inheritance in M-LOOP, because the responsibility of the base class is quite tightly scoped and well-defined. I suppose it's just that seeing another class inserted into the hierarchy is another reminder to me of the way things are. While I have an instinctive negative reaction to that, I think it's still a good solution given the context. 

All of that is to say, just ignore me, and maybe if you have more consolidation planned we might be able to think of a way to rework things a bit more significantly.

**Comment by zakv** [comment]
I think I see what you're getting at. If I understand correctly, in your experience the "[composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance)" philosophy tends to work better, especially when the API/expectations of the composite classes are well laid out.

Something like that does sound much easier to maintain and extend. I agree that M-LOOP is pretty primed for a major rewrite to implement large changes like that. Admittedly though that's probably too daunting of a task to be worth anyone's time at the moment, especially given that there aren't any major new features planned and, as you say, things generally work as is.

**Pull request merged**